pkg/hooks/audit_logger_sqlite.go

```diff
@@ -0,0 +1,184 @@
+
+type SQLiteAuditLogger struct {
+	sqlite *sql.DB
+	logger *logging.Logger
+}
+
+func NewSQLiteAuditLogger(sqliteDBPath string, logger *logging.Logger) (*SQLiteAuditLogger, error) {
+	db, err := sql.Open("sqlite3", sqliteDBPath)
+	if err != nil {
+		return nil, util.Errorf("Unable to open sqlite at %s. %v", sqliteDBPath, err)
+	}
+	if db == nil {
+		return nil, util.Errorf("Unable to create a sqlite connection.")
+	}
+	al := &SQLiteAuditLogger{
+		sqlite: db,
+		logger: logger,
+	}
+	if err := al.ensureMigrated(); err != nil {
+		return nil, util.Errorf("Failed to apply migrations to sqlite DB: %v", err)
+	}
+	return al, nil
+}
+
+func (al *SQLiteAuditLogger) log(ctx *HookExecContext, success bool) {
+	env := ctx.env
+	stmt := `INSERT INTO hook_results(
+pod_id,
+pod_unique_key,
+hook_name,
+hook_stage,
+success
+) VALUES(?,?,?,?,?)`
+	podID := env.HookedPodIDEnvVar
+	podUniqueKey := env.HookedPodUniqueKeyEnvVar
+	hookName := ctx.Name
+	hookStage := env.HookEventEnvVar
+	// Boolean values are represented by TINYINTs in SQL. Adapt.
+	dbSuccess := 0
+	if !success {
+		dbSuccess = 1
+	}
+
+	err := al.withRetries(func() error {
+		_, err := al.sqlite.Exec(stmt, podID, podUniqueKey, hookName, hookStage, dbSuccess)
+		return err
+	}, 3)
+	if err != nil {
+		al.logger.WithError(err).Errorln("error executing log statement")
+	}
+	al.trimStaleRecords()
+}
+
+func (al *SQLiteAuditLogger) trimStaleRecords() {
+	stmt := fmt.Sprintf(`DELETE FROM hook_results WHERE date <= (SELECT(datetime('now', '-%d seconds')))`, *auditLogTTL)
+
+	err := al.withRetries(func() error {
+		_, err := al.sqlite.Exec(stmt)
+		return err
+	}, 3)
+	if err != nil {
+		al.logger.WithError(err).Errorln("Error executing cleanup routine!")
+	}
+}
+
+func (al *SQLiteAuditLogger) LogSuccess(ctx *HookExecContext) {
+	al.log(ctx, true)
+}
+
+func (al *SQLiteAuditLogger) LogFailure(ctx *HookExecContext, err error) {
+	al.log(ctx, false)
+}
+
+var (
+	sqliteMigrations = []string{
+		`create table hook_results (
+	      id integer not null primary key autoincrement,
+	      date datetime default current_timestamp,
+	      pod_id text,
+	      pod_unique_key text,
+	      hook_name string,
+	      hook_stage string,
+	      success tinyint);`,
+
+		"create index hook_results_hook_name on hook_results(hook_name);",
+		"create index hook_results_pod_id on hook_results(pod_id);",
+		// FUTURE MIGRATIONS GO HERE
+	}
+)
+
+const (
+	sqliteCreateSchemaVersionTable = `create table if not exists hooks_schema_version ( version integer );`
+	getSchemaVersionQuery          = `select version from hooks_schema_version;`
+	updateSchemaVersionStatement   = `update hooks_schema_version set version = ?;`
+)
+
+// Close will terminate this AuditLogger. Re-establishing the connection is not supported, use the constructor.
+func (al *SQLiteAuditLogger) Close() error {
+	return al.sqlite.Close()
+}
+
+func (al *SQLiteAuditLogger) withRetries(f func() error, n int) error {
+	var err error
+	err = f()
```

Comment:
I see Github thinks https://github.com/square/p2/pull/823#discussion_r113801665 is out of date but I say it still applies!